File: Samples/blazor-add-in/excel-blazor-add-in/excel-blazor-add-in/Pages/Index.razor.js
/* Copyright(c) Maarten van Stam. All rights reserved. Licensed under the MIT License. */
/**
 * Basic function to show how to insert a value into cell A1 on the selected Excel worksheet.
 */
export function helloButton() {

    return Excel.run(context => {

        // Insert text 'Hello world!' into cell A1.
        context.workbook.worksheets.getActiveWorksheet().getRange("A1").values = [['Hello world!']];

        // sync the context to run the previous API call, and return.
        return context.sync();
    });
}

export function sheetValues() {
    return Excel.run(function (context) {
        const sheet = context.workbook.worksheets.getActiveWorksheet();
        const usedRange = sheet.getUsedRange();
        usedRange.load("values");

        return context.sync().then(function () {
            return usedRange.values;
        });
    }).catch(function (error) {
        console.log("Error: " + error);
        if (error instanceof OfficeExtension.Error) {
            console.log("Debug info: " + JSON.stringify(error.debugInfo));
        }
    });
}

export function formulaOfSelected() {
    return Excel.run(function (context) {
        const selectedRange = context.workbook.getSelectedRange();
        const cellToAnalyse = selectedRange.getCell(0, 0);

        // Load the formula of the selected cell
        cellToAnalyse.load("formulas");

        return context.sync().then(function () {
            return cellToAnalyse.formulas;
        });
    }).catch(function (error) {
        console.log("Error: " + error);
        if (error instanceof OfficeExtension.Error) {
            console.log("Debug info: " + JSON.stringify(error.debugInfo));
        }
    });
}

File: Samples/blazor-add-in/excel-blazor-add-in/excel-blazor-add-in/Pages/Index.razor.cs
/* Copyright(c) Maarten van Stam. All rights reserved. Licensed under the MIT License. */

using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;

namespace BlazorAddIn.Pages
{
    /// <summary>
    /// Starter class to demo how to insert `Hello world!` text.
    /// </summary>
    public partial class Index
    {
        [Inject]
        public IJSRuntime JSRuntime { get; set; } = default!;
        public IJSObjectReference JSModule { get; set; } = default!;

        protected override async Task OnAfterRenderAsync(bool firstRender)
        {
            if (firstRender)
            {
                JSModule = await JSRuntime.InvokeAsync<IJSObjectReference>("import", "./Pages/Index.razor.js");
            }
        }

        /// <summary>
        /// Basic function to invoke inserting `Hello world!` text.
        /// </summary>
        /// 

        private async Task HelloButton()
        {
            await JSModule.InvokeVoidAsync("helloButton");

            //get all the values from the active sheet
            var sheetValues = await JSModule.InvokeAsync<string>("sheetValues");

            //get the formula of the active cell
            var cellFormula = await JSModule.InvokeAsync<object>("formulaOfSelected");
            System.Diagnostics.Debug.WriteLine(cellFormula);

            //do some processing here

            //write the value back to the Office Addin window
        }
    }
}
